### PR TITLE
Refactor and merge Callbacks

### PR DESCRIPTION
**Pull request by gerardsn**
cleanup and merge callbacks. This contains some fixes listed in #3152, but the majority is left for when we have access to a compliant issuer so we can do an end to end test of the flow.

### CODE REVIEW

File: auth/api/iam/api.go [hunk 1]
```diff
@@ -208,19 +208,52 @@ func (r Wrapper) HandleTokenRequest(ctx context.Context, request HandleTokenRequ
...
+	// check did in path
+	ownDID, err := r.toOwnedDID(ctx, request.Did)
 	if err != nil {
-		// this is an OAuthError already, will be rendered as 400 but that's fine (for now) for an illegal id
 		return nil, err
 	}
+	// check if state is present and resolves to a client state
+	if request.Params.State == nil || *request.Params.State == "" {
+		return nil, oauthError(oauth.InvalidRequest, "missing state parameter")
+	}
```
Comment (woutslakhorst) [inline auth/api/iam/api.go:220]: If the Callback has an error but is missing the client state, the actual error will not be shown

File: auth/api/iam/api.go [hunk 2]
```diff
@@ -208,19 +208,52 @@ func (r Wrapper) HandleTokenRequest(ctx context.Context, request HandleTokenRequ
...
 
-	return r.handleCallback(ctx, request)
+	// check if code is present
+	if request.Params.Code == nil || *request.Params.Code == "" {
+		return nil, withCallbackURI(oauthError(oauth.InvalidRequest, "missing code parameter"), oauthSession.redirectURI())
+	}
+
+	// continue flow
+	switch oauthSession.ClientFlow {
+	case "openid4vci_credential_request":
```
Comment (woutslakhorst) [inline auth/api/iam/api.go:249]: make constants

File: auth/api/iam/openid4vci.go [hunk 3]
```diff
@@ -79,24 +83,21 @@ func (r Wrapper) RequestOid4vciCredentialIssuance(ctx context.Context, request R
...
+	redirectUri = redirectUri.JoinPath(oauth.CallbackPath)
 	// Store the session
-	err = r.openid4vciSessionStore().Put(state, &Oid4vciSession{
-		HolderDid:         requestHolder,
-		IssuerDid:         issuerDid,
-		RemoteRedirectUri: request.Body.RedirectUri,
-		RedirectUri:       redirectUri.String(),
-		PKCEParams:        pkceParams,
+	err = r.oauthClientStateStore().Put(state, &OAuthSession{
+		ClientFlow:  "openid4vci_credential_request",
```
Comment (woutslakhorst) [inline auth/api/iam/openid4vci.go:93]: const

File: docs/_static/auth/iam.partial.yaml [hunk 4]
```diff
@@ -472,64 +472,6 @@ paths:
                 "$ref": "#/components/schemas/VerifiableCredential"
         default:
           $ref: '../common/error_response.yaml'
-  /iam/oid4vci/callback:
```
Comment (woutslakhorst) [inline docs/_static/auth/iam.partial.yaml:475]: also add to calback API description that it's used for both calls.

**Review by woutslakhorst — CHANGES_REQUESTED** [review]

File: auth/api/iam/api.go (hunk 1 above)
Comment (gerardsn) [reply auth/api/iam/api.go:220]: RFC6749 [4.1.2.1](https://www.rfc-editor.org/rfc/rfc6749.html#section-4.1.2.1).  Error Response
> state
         REQUIRED if a "state" parameter was present in the client
         authorization request.  The exact value received from the
         client.

I wasn't sure what to do with this, but I will wrap the error if there is one

**Review by woutslakhorst — APPROVED** [review]

**Pull request merged**